### PR TITLE
Shutdown HttpApp without STDIN #914

### PR DESCRIPTION
**Pull request by shishkin**
* Prevent immediate shutdown when STDIN is not present, e.g. under sbt-revolver;
* Shutdown the server upon JVM termination;

Couldn't find immediate way of writing a test for the change without introducing indirection of JVM shutdown hook and `StdIn`. Does anyone have any suggestions? I obviously verified the expected behavior in both sbt-revolver (shutdown via C-c) and console (via ENTER).

I will update the `javadsl` implementation as well if this is going in the right direction.

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @shishkin,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by jlprat** [comment]
OK TO TEST

**Comment by jlprat** [comment]
OK TO TEST

**Comment by jlprat** [comment]
OK TO TEST

**Comment by ktoso** [comment]
PLS BUILD

File: akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala [hunk 1]
```diff
@@ -132,10 +132,15 @@ abstract class HttpApp extends Directives {
...
    */
   protected def waitForShutdownSignal(system: ActorSystem)(implicit ec: ExecutionContext): Future[Done] = {
+    val promise = Promise[Done]()
+    sys.addShutdownHook {
+      promise.success(Done)
+    }
     Future {
-      StdIn.readLine("Press RETURN to stop...\n")
-      Done
+      if (StdIn.readLine("Press RETURN to stop...\n") != null)
```
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: Do you really need the if in this case? `StdIn.readline` is blocking anyway and no matter what it returns we want the promise to complete.

File: akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java [hunk 2]
```diff
@@ -145,15 +145,18 @@ protected void postHttpBindingFailure(Throwable cause) {
...
    */
   protected CompletionStage<Done> waitForShutdownSignal (ActorSystem system) {
-    return CompletableFuture.supplyAsync( () -> {
+    final CompletableFuture<Done> promise = new CompletableFuture<>();
+    Runtime.getRuntime().addShutdownHook(new Thread(() -> promise.complete(Done.getInstance())));
+    CompletableFuture.runAsync(() -> {
       System.out.println("Press RETURN to stop...");
       try {
-        System.in.read();
+        if (System.in.read() >= 0)
```
Comment (jlprat) [inline akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java]: Do you really need the if in this case? `System.in.read` is blocking anyway and no matter what it returns we want the promise to complete.

File: akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala (hunk 1 above)
Comment (shishkin) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: In my testing with sbt-revolver is returns `null` when STDIN is not available.

File: akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java (hunk 2 above)
Comment (shishkin) [inline akka-http/src/main/java/akka/http/javadsl/server/HttpApp.java]: In my testing with sbt-revolver is returns `-1` when STDIN is not available.

**Comment by jlprat** [comment]
I have some minor questions about it, but it looks good! :smiley:

File: akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala (hunk 1 above)
Comment (jlprat) [inline akka-http/src/main/scala/akka/http/scaladsl/server/HttpApp.scala]: OK, that makes sense :)

**Comment by shishkin** [comment]
Thanks @jlprat

**Comment by jrudolph** [comment]
I tried a quick test app and it works for me.

LGTM

**Pull request merged**